### PR TITLE
RWA-1126 Auto-assign task to case role

### PR DESCRIPTION
**Pull request by mkayad**
[RWA-1126](https://tools.hmcts.net/jira/browse/RWA-1126)

### JIRA link (if applicable) ###

[RWA-1126](https://tools.hmcts.net/jira/browse/RWA-1126)

### Change description ###

See [RWA-1126](https://tools.hmcts.net/jira/browse/RWA-1126)

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 1]
```diff
@@ -119,14 +125,22 @@ public boolean checkAssigneeIsStillValid(TaskResource taskResource, String assig
...
+                } else if (taskRoleResource.getAutoAssignable()
+                            && taskRoleResource.getAuthorizations() != null
+                            && taskRoleResource.getAuthorizations().length == 0) {
+                    return true;
+                } else if (taskRoleResource.getAutoAssignable()
+                           && taskRoleResource.getAuthorizations() == null) {
+                    return true;
                 } else {
                     return findMatchingRoleAssignment(taskRoleResource, roleAssignment);
                 }
```
Comment (alexbaeza) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:146]: I would look into trying to combine some of these if statements as others in terms of performance you would be accessing the same value multiple times which is not a major issue and sometimes for readability could be okay.

The other thing I would suggest is maybe wrapping this filtering in its own method so it reduces the complexity and can make it easier to maintain and read

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 2]
```diff
@@ -119,14 +125,22 @@ public boolean checkAssigneeIsStillValid(TaskResource taskResource, String assig
         List<RoleAssignment> orderedRoleAssignments = orderRoleAssignments(rolesList, roleAssignments);
         return orderedRoleAssignments.stream()
             .filter(roleAssignment -> {
```
Comment (alexbaeza) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:127]: For extra safety we could also add add an extra  to check for roleAssignment.getRoleName() not empty
```

```

Comment (mkayad) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:127]: I checked this scenario and added a test for it, this method orderRoleAssignments is filtering out any role assignment with a null or empty rolename

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java (hunk 1 above)
Comment (mkayad) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java]: That is done

**Review by serdaroguzhanakin — APPROVED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java [hunk 3]
```diff
@@ -496,11 +496,11 @@ void should_return_201_with_task_assigned() throws Exception {
                 jsonPath("$.task_id").value(taskId),
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java:496]: can we rename test as state is now UNASSIGNED?

**Review by BenNzewi — COMMENTED** [review]
Do we have an FT for autoAssign?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java (hunk 3 above)
Comment (mkayad) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java:496]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java [hunk 4]
```diff
@@ -757,12 +757,11 @@ void should_return_201_with_task_assigned_when_invalid_old_assignee_auto_assign_
...
                 jsonPath("$.task_name").value("aTaskName"),
                 jsonPath("$.task_type").value("followUpOverdueReasonsForAppeal"),
-                jsonPath("$.state").value("ASSIGNED"),
```
[the remark this answers is not in the document]
Comment (mkayad) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java:760]: done

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 5]
```diff
@@ -118,27 +118,57 @@ public boolean checkAssigneeIsStillValid(TaskResource taskResource, String assig
...
+            return false;
+        } else {
+            return findMatchingRoleAssignment(taskRoleResource, roleAssignment);
+        }
+    }
+
+    private boolean isTaskRoleNotAutoAssignableOrAuthorisationsNotMatching(RoleAssignment roleAssignment,
+                                                                           TaskRoleResource taskRoleResource) {
+        return !taskRoleResource.getAutoAssignable()
+               || taskRoleResource.getAutoAssignable()
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:142]: | taskRoleResource.getAutoAssignable()
                  && isAuthorisationsValid(taskRoleResource)

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 6]
```diff
@@ -118,27 +118,57 @@ public boolean checkAssigneeIsStillValid(TaskResource taskResource, String assig
...
+    private boolean isTaskRoleAutoAssignableWithNullOrEmptyAuthorisations(TaskRoleResource taskRoleResource) {
+        return taskRoleResource.getAutoAssignable()
+               && (!isAuthorisationsValid(taskRoleResource)
+                   || taskRoleResource.getAuthorizations() == null);
+    }
+
+    private boolean hasTaskBeenAssigned(TaskResource taskResource, TaskRoleResource taskRoleResource) {
+        return !taskRoleResource.getAutoAssignable()
+            && taskResource.getAssignee() != null
+            && isAuthorisationsValid(taskRoleResource);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:156]: unit test to cover?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 7]
```diff
@@ -118,27 +118,57 @@ public boolean checkAssigneeIsStillValid(TaskResource taskResource, String assig
...
+        return !taskRoleResource.getAutoAssignable()
+               || taskRoleResource.getAutoAssignable()
+                  && isAuthorisationsValid(taskRoleResource)
+                  && !findMatchingRoleAssignment(taskRoleResource, roleAssignment);
+    }
+
+    private boolean isTaskRoleAutoAssignableWithNullOrEmptyAuthorisations(TaskRoleResource taskRoleResource) {
+        return taskRoleResource.getAutoAssignable()
+               && (!isAuthorisationsValid(taskRoleResource)
+                   || taskRoleResource.getAuthorizations() == null);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:150]: unit test to cover?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 8]
```diff
@@ -118,27 +118,57 @@ public boolean checkAssigneeIsStillValid(TaskResource taskResource, String assig
...
+    }
+
+    private boolean hasTaskBeenAssigned(TaskResource taskResource, TaskRoleResource taskRoleResource) {
+        return !taskRoleResource.getAutoAssignable()
+            && taskResource.getAssignee() != null
+            && isAuthorisationsValid(taskRoleResource);
+    }
+
+    private boolean isAuthorisationsValid(TaskRoleResource taskRoleResource) {
+        return  taskRoleResource.getAuthorizations() != null
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:160]: unit test to cover?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java [hunk 9]
```diff
@@ -757,11 +756,12 @@ void should_return_201_with_task_unassigned_when_invalid_old_assignee_auto_assig
...
                 jsonPath("$.task_name").value("aTaskName"),
                 jsonPath("$.task_type").value("followUpOverdueReasonsForAppeal"),
-                jsonPath("$.state").value("UNASSIGNED"),
+                jsonPath("$.state").value("ASSIGNED"),
+                jsonPath("$.assignee").value("anotherAssignee"),
                 jsonPath("$.task_system").value("SELF"),
                 jsonPath("$.security_classification").value("PUBLIC"),
                 jsonPath("$.title").value("aTaskName"),
-                jsonPath("$.auto_assigned").value(false),
+                jsonPath("$.auto_assigned").value(true),
```
Comment (mkayad) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java:768]: I would expect the auto assignable to be false when the auto assignable flag is false for the role.

**Review by mkayad — COMMENTED** [review]
These 2 files were deleted, not sure why

**Review by mkayad — COMMENTED** [review]
LGTM

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java (hunk 9 above)
Comment (dharmendrak) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java]: reverted

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetJudicialTaskControllerCFTTest.java [hunk 10]
```diff
@@ -36,19 +35,15 @@
     @Before
     public void setUp() {
         authenticationHeaders = authorizationProvider.getJudgeAuthorization("wa-ft-test-r2-");
-
     }
 
     @Test
-    public void should_return_a_200_with_allocateHearingJudge_task_and_standard_grant_type() {
+    public void should_return_a_200_with_reviewHearingBundle_task_and_standard_grant_type() {
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetJudicialTaskControllerCFTTest.java:41]: Nice. 👍

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetJudicialTaskControllerCFTTest.java [hunk 11]
```diff
@@ -77,26 +72,26 @@ public void should_return_a_200_with_allocateHearingJudge_task_and_standard_gran
...
         ZonedDateTime createdDate = ZonedDateTime.now();
         String formattedCreatedDate = CAMUNDA_DATA_TIME_FORMATTER.format(createdDate);
         ZonedDateTime dueDate = createdDate.plusDays(1);
         String formattedDueDate = CAMUNDA_DATA_TIME_FORMATTER.format(dueDate);
 
         InitiateTaskRequest req = new InitiateTaskRequest(INITIATION, asList(
-            new TaskAttribute(TASK_TYPE, "allocateHearingJudge"),
-            new TaskAttribute(TASK_NAME, "Allocate hearing judge"),
+            new TaskAttribute(TASK_TYPE, taskType),
+            new TaskAttribute(TASK_NAME, taskName),
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetJudicialTaskControllerCFTTest.java:94]: Can we include IT's to cover the other 2 AC's?

1. If auto-assignable is true + authorisation is empty = assign task
2. If auto-assignable is true + authorisation has value -> check authorisation, if no match = do not assign task

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerCFTTest.java [hunk 12]
```diff
@@ -459,7 +459,7 @@ public void should_return_a_200_with_task_description_property() {
             .body("task.case_type_id", notNullValue())
             .body("task.case_category", notNullValue())
             .body("task.case_name", notNullValue())
-            .body("task.auto_assigned", notNullValue())
+            .body("task.auto_assigned", equalTo(true))
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerCFTTest.java:462]: if this FT is covering "If auto-assignable is true + authorisation is empty = assign task" shouldn't the **task.task_state = "assigned"**

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java [hunk 13]
```diff
@@ -674,7 +675,7 @@ void should_return_201_with_task_assigned_when_valid_old_assignee() throws Excep
     }
 
     @Test
-    void should_return_201_with_task_assigned_when_invalid_old_assignee_auto_assign_case() throws Exception {
+    void should_return_201_with_task_unassigned_when_invalid_old_assignee_auto_assign_case() throws Exception {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/PostInitiateByIdControllerTest.java:678]: This does not cover this AC -> If auto-assignable is true + authorisation is empty = assign task. Te test set state to UNASSIGNED + auto_assigned is false

**Review by BenNzewi — APPROVED** [review]

**Pull request merged**